apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java

```diff
@@ -160,10 +157,15 @@ String fetchConfig(final ConfigurationRegistry configurationRegistry) {
             return null;
         }
         try {
+            payloadSerializer.blockUntilReady();
```

Comment:
Not saying this doesn't work but did you consider providing a `Future<MetaData>` in the constructor and blocking on `Future#get` here and call `DslJsonSerializer#serializeMetadata(MetaData metaData)`? That way, we could keep state management out of the `DslJsonSerializer`.

Reply:
Of course, this is how I started, but there were several reasons I changed that: the first is that nobody really needs the metadata object, it is spread it around just so it can be handed to serialization. 
In addition, I wanted to create a central serialization for it, meaning that `MetaData` (or something else) needs to serialize and hold the serialized data. This would mean that `MetaData` (or something else) now either create and use a `PayloadSerializer`, or I extract the metadata serialization from the rest, but that proved to be too painful and risky for this PR. Both options looked bad when I tried them.
Lastly, I don't think it makes sense to block during serialization itself, for example- you better only start the HTTP request when ready to write to it. So I thought it would be better if the serialization interface provides a good indication for that, instead of relying on `Future#get` (this is not the only usage of it).

Bottom line - there are many options to do that and I played with several ending with this. Given the multiple dependencies we had on `MetaData` and the time I wanted to spend on this, this seems as the least ugly.

Reply:
Thanks for explaining the background

Reply:
I don't mind, but I think it is weird to explain why something was chosen over something else you wouldn't necessarily consider as a default option. Imagine you only see this code now for the first time- `MetaData` can be easily an internal serialization thing not known to the outer world (it is not package private only not to broaden the refactoring of this PR even more). In such case, this is the interface of the serializer- it needs time to fully initialize, no matter why, so users of it should be aware. I made efforts to emphasize this in the API as much as possible, including javadoc and an Exception which is part of the API.